Implement Locking System Across Multiple Pavilion Instances

hwikle-lanl opened the pull request:
Code review checklist:

- [ ] Code is generally sensical and well commented
- [ ] Variable/function names all telegraph their purpose and contents
- [ ] Functions/classes have useful doc strings
- [ ] Function arguments are typed
- [ ] Added/modified unit tests to cover changes.
- [ ] New features have documentation added to the docs.
- [ ] New features and backwards compatibility breaks are noted in the RELEASE.md

Commit:
add notes

Commit:
start towards prototype solution

Commit:
track build hashes on MultiBuildTracker

Commit:
restore lost changes

Commit:
fix mistake

Commit:
implement hash->lock map

Commit:
fix lingering misnamed function call

hwikle-lanl committed:
remove stray code

hwikle-lanl committed:
cache build hash on creation

hwikle-lanl committed:
attempt to resolve style test failure

hwikle-lanl committed:
second attempt at style test

hwikle-lanl committed:
add "fuzzy" lockfile implementation

hwikle-lanl committed:
fix minor style issue

hwikle-lanl committed:
reimplement timeout on build locks

hwikle-lanl committed:
update build locking test

hwikle-lanl committed:
ignore vscode workspace

hwikle-lanl committed:
attempt to fix timeout

hwikle-lanl committed:
fix timed lock context management

hwikle-lanl committed:
fix style issues; tweak lock context

hwikle-lanl committed:
fix mistake; more style fixes

hwikle-lanl committed:
fix changed method name

hwikle-lanl committed:
another style fix

hwikle-lanl committed:
even more style fixes

hwikle-lanl committed:
(attempt to) fix broken unit test

hwikle-lanl committed:
fix mistake with protected class

hwikle-lanl committed:
fix accidentally deleted import

hwikle-lanl committed:
restore removed import

hwikle-lanl committed:
attempt to fix broken docstring

hwikle-lanl committed:
restore (janky) docstring to pass unit test

hwikle-lanl committed:
fix unit test

hwikle-lanl committed:
fix unit test

hwikle-lanl committed:
fix unit test

hwikle-lanl committed:
fix unit test

hwikle-lanl committed:
Merge branch 'fuzzy-lock' into build-race-condition

hwikle-lanl committed:
tweak status file check

hwikle-lanl committed:
style fixes

hwikle-lanl committed:
fix build lock unit test

hwikle-lanl committed:
remove fuzzy lock -- for inclusion later

hwikle-lanl committed:
Merge remote-tracking branch 'origin/build-race-condition' into build-race-condition

Resolve bad repo state

hwikle-lanl committed:
add sphinx doctring; minor clean up

hwikle-lanl committed:
edit docstrings

hwikle-lanl committed:
integrate NFSLock into builder and mb_tracker

hwikle-lanl committed:
progress towards integrating NFSLock object

hwikle-lanl committed:
fix some failing tests

hwikle-lanl committed:
pass test_rebuild test

hwikle-lanl committed:
Merge branch 'master' into fuzzy-lock

hwikle-lanl committed:
fix style issues

hwikle-lanl committed:
fix NameError

hwikle-lanl committed:
implement PR feedback

hwikle-lanl committed:
fix style issues

hwikle-lanl committed:
remove lockfile directory once empty

hwikle-lanl committed:
miscellaneous changes

hwikle-lanl committed:
remove use of lock timeout in builder

hwikle-lanl committed:
implement code review feedback

hwikle-lanl committed:
add tests for fuzzylock

hwikle-lanl committed:
return lockfile from __enter__ method

hwikle-lanl committed:
create tests for FuzzyLock

hwikle-lanl committed:
pass fuzzylock mutual exclusion test

hwikle-lanl committed:
fix FuzzyLock cleanup test

hwikle-lanl committed:
fix style issues

hwikle-lanl committed:
remove debug print statements

hwikle-lanl committed:
more style fixes

Paul-Ferrell commented on lib/pavilion/lockfile.py:
```diff
@@ -328,3 +329,99 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+
+        :return: True if lock has been successfully acquired; False if timed out
+        """
+        self._lock_dir.mkdir(exist_ok=True)
+
+        # Declare intent to take lock
+        self._lockfile.touch(exist_ok=False)
```
Catch and reraise as a custom error.  See lib/pavilion/errors.py

Paul-Ferrell commented on lib/pavilion/lockfile.py:
```diff
@@ -328,3 +329,99 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+        del self._mtimes[self._lockfile]
+
+        # Remove lock directory once it's empty
+        if len(list(self._lock_dir.iterdir())) == 0:
```
Catch a generic OSError.  No need to check if directory is empty before trying to delete it.

Paul-Ferrell commented on lib/pavilion/lockfile.py:
```diff
@@ -328,3 +329,99 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+        :return: Path object to whichever lockfile was created first"""
+
+        lfiles = list(self._lock_dir.iterdir())
+        dict_keys = list(self._mtimes.keys())
```
Just iterate over the list directly.

Paul-Ferrell commented on test/tests/lock_tests.py:
```diff
@@ -136,3 +145,76 @@ def _acquire_lock(*args, **kwargs):
+        def check_file(idx, results):
+            file = None
+
+            with lockfile.FuzzyLock(self.lock_dir) as lock:
```
Add a very short sleep.

hwikle-lanl committed:
implement code review feedback

hwikle-lanl committed:
fix failing tests

Paul-Ferrell commented on lib/pavilion/lockfile.py:
```diff
@@ -328,3 +330,113 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+        :return: True if lock has been successfully acquired; False if timed out
+        """
+
+        try:
+            self._lock_dir.mkdir(exist_ok=True)
+
+            # Declare intent to take lock
+            self._lockfile.touch(exist_ok=False)
+        except OSError:
+            msg = f"Unable to create lockfile {self._lockfile}."
+
+            if not self._lock_dir.exists():
+                # This is probably the fault of the caller
+                msg += " Lock directory ({self._lock_dir}) does not exist."
+            elif self._lockfile.exists():
+                # If this happens, something went very wrong. This is likely dead code.
+                msg += " Lockfile ({self._lock_dir}) already exists."
+
+            raise LockFileError(msg)
```
One try/except per thing that could throw an exception.

Paul-Ferrell commented on lib/pavilion/lockfile.py:
```diff
@@ -328,3 +330,113 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                    pass
+
+        # Sort files by creation time and return oldest
+        return sorted(list(self._mtimes.items()), key=lambda x: x[1])[0][0]
```
No need to cast to list.

hwikle-lanl committed:
Implement PR feedback

Separate different modes of failure into separate try/catch sections,
for lockfile creation, and include information about previous errors.
Also remove unnecessary cast to list when sorting lockfiles and
correct outdated docstring for FuzzyLock __enter__ method.

Paul-Ferrell committed:
Merge branch 'develop' into fuzzy-lock

Paul-Ferrell committed:
Merge branch 'develop' into fuzzy-lock

Paul-Ferrell committed:
Merge branch 'develop' into fuzzy-lock
